File: code/operating_system/src/scheduling/job_sequencing/job_sequencing.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Job {
    char id;
    int deadline;
    int profit;
};

bool jobComparator(const Job &a, const Job &b) {
    return a.profit > b.profit;
}

void jobSequence(vector<Job> &jobs) {
    int n = jobs.size();
    
    sort(jobs.begin(), jobs.end(), jobComparator);
    
    int maxDeadline = 0;
    for (const Job &job : jobs) {
        maxDeadline = max(maxDeadline, job.deadline);
    }
    
    vector<bool> slot(maxDeadline, false);
    
    vector<char> sequence;
    int totalProfit = 0;
    
    for (const Job &job : jobs) {
        for (int i = min(maxDeadline - 1, job.deadline - 1); i >= 0; i--) {
            if (!slot[i]) {
                slot[i] = true;
                sequence.push_back(job.id);
                totalProfit += job.profit;
                break;
            }
        }
    }
    
    cout << "Job sequence: ";
    for (char jobId : sequence) {
        cout << jobId << " ";
    }
    cout << "\nTotal profit: " << totalProfit << endl;
}

int main() {
    vector<Job> jobs = {
        {'a', 2, 100},
        {'b', 1, 19},
        {'c', 2, 27},
        {'d', 1, 25},
        {'e', 3, 15}
    };
    
    jobSequence(jobs);
    
    return 0;
}

File: code/job_sequencing.java
import java.util.*;

class Job {
    char id;
    int deadline;
    int profit;

    public Job(char id, int deadline, int profit) {
        this.id = id;
        this.deadline = deadline;
        this.profit = profit;
    }
}

public class JobSequencing {
    public static void main(String[] args) {
        List<Job> jobs = new ArrayList<>();
        jobs.add(new Job('a', 2, 100));
        jobs.add(new Job('b', 1, 19));
        jobs.add(new Job('c', 2, 27));
        jobs.add(new Job('d', 1, 25));
        jobs.add(new Job('e', 3, 15));

        // Sort the jobs in decreasing order of profit
        jobs.sort((j1, j2) -> Integer.compare(j2.profit, j1.profit));

        int maxDeadline = 0;
        for (Job job : jobs) {
            if (job.deadline > maxDeadline) {
                maxDeadline = job.deadline;
            }
        }

        char[] result = new char[maxDeadline];
        Arrays.fill(result, ' ');

        int totalProfit = 0;

        for (Job job : jobs) {
            for (int i = job.deadline - 1; i >= 0; i--) {
                if (result[i] == ' ') {
                    result[i] = job.id;
                    totalProfit += job.profit;
                    break;
                }
            }
        }

        System.out.println("Maximized Profit: " + totalProfit);
        System.out.println("Job Sequence: " + new String(result));
    }
}
